doc: dev_process: Add description about GH labels usage

aescolar opened the pull request:
There has been some confusion about how GitHub labels should be used.
This commit adds a new doc page, which tries to clarify this.
This discussion started in this doc:
https://docs.google.com/document/d/1EnWOJERsdygOMnymlhKbXkuA5mGPhd9OyX9htJBKXMk/edit

UPDATE: convenience bookmark: https://docs.zephyrproject.org/latest/development_process/dev_env_and_tools.html#labeling-issues-and-pull-requests-in-github
Follow-up in issue #15156 

--------------

Once this PR is merged I will update the labels descriptions in GH to align them with the ones in this page.
Moreover I will delete the following labels (unless somebody disagrees):
```
CII: Silver
won’t fix   (Default GH label) instead describe in a comment why, and close the issue.
invalid               	“                          	”
task
todo
Revisit     Seems inconsistent, better to use milestone tags to indicate when to revisit (?)
TEE         Duplicate of “area: Trusted execution environment” (?)
Process     Duplicate of "area:Process" (?)
```
I will also rename labels to start with uppercase consistently
~~Will rename Trivial => "Trivial PR" (To avoid developers assigning it to trivial bugs.~~

aescolar committed:
doc: dev_process: Add description about GH labels usage

There has been some confusion about how GitHub labels
should be used.
This commit adds a new doc page, which tries to clarify this.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
+Description    Indicates which areas are affected by the bug or pull request
+=============  ===============================================================
+
+An area maintainer should be able to filter by his area and
```
```suggestion
An area maintainer should be able to filter by an area label and
```

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
+=============  ===============================================================
+Labels         ``priority:{high|medium|low}``
+Applicable to  Issues only
+Description    To classify the priority/importance of a bug or feature
```
```suggestion
Description    To classify the impact and importance of a bug or feature
```

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
+Description    To classify the priority/importance of a bug or feature
+=============  ===============================================================
+
+Note: Issue priorities are decided/changed during the bug-triage or TSC
```
```suggestion
Note: Issue priorities are generally set or changed during the issue triage or TSC.
```

aescolar replied:
I take the 1st change. About the 2nd, the meeting invite actually calls it "Bug Triage & Release Readiness", so I was trying to be consistent with that.

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
++-------------------+----------------------------------------------------------+
+|``Coverity``       |A Coverity detected issue or its fix                      |
++-------------------+----------------------------------------------------------+
+|Waiting for        | The Zephyr developers are waiting for the submitter to   |
```
This label is currently called ``waiting for response``.  This name should also be enclosed in backticks to be formatted like all the others in this table.

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
+                 For work in progress, GitHub "draft" PRs are preferred
+``Stale PR``     PR which seems abandoned, and requires attention by the author
+``Blocked``      Blocked by another PR or issue
+``needs review`` The PR needs attention from the maintainers
```
Is this an opportunity to have all labels with a consistent upper-case first letter?

aescolar replied:
I can certainly fix that :)

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
+=================
+
+==================== ===========================================================
+``Regression``       Something, which was working, does not anymore
```
```suggestion
``Regression``       Something, which was working, but does not anymore
```

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
+==================== ===========================================================
+``Regression``       Something, which was working, does not anymore
+                     (bug subtype)
+``question``         This issue is a question to the Zephyr developers
```
A consistent first upper-case letter would be good...

Also, are we encouraging folks to ask questions as GitHub issues, or asking them to post questions in Slack?

aescolar replied:
>A consistent first upper-case letter would be good..

Sure

> Also, are we encouraging folks to ask questions as GitHub issues, or asking them to post questions in Slack?

Well that is another thing. The fact is that today, people comes to both places with questions, and some issues are morphed into questions.

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,128 @@
+
+All GitHub issues or pull requests must be appropriately labeled.
+When reviewing a PR, if it has missing or incorrect labels, maintainers shall
+fix it.
```
Maybe add:
```
Issues and PRs often have multiple labels assigned,
to help classify them in the different available categories.
```

aescolar commented:
@dbkinder Like always, thanks David. Your comments should be addressed now (except for what was pointed above) Please take another look and tell if you find anything else.

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,130 @@
+=============  ===============================================================
+Labels         ``Area:*``
+Applicable to  PRs  and issues
+Description    Indicates which areas are affected by the bug or pull request
```
I thought I had added a comment about this, but doesn't look like I saved it.  Instead of using the word "area" to define what an "area" is, how about something like:
```
Indicates subsystems (e.g., Kernel, I2C, Memory Management),
project functions (e.g., Debugging, Documentation, Process), or other
categories (e.g., Coding Style, MISRA-C)  affected by the bug or pull request.
```

aescolar replied:
Done :)

dbkinder reviewed (changes requested):
Looking better.  I left off a comment on the previous review...

dbkinder reviewed (approved):
LGTM, thanks!

carlescufi commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,133 @@
+=============  ===============================================================
+Labels         ``API``, ``dev-review``, ``TSC``
+Applicable to  PRs  and issues
+Description    The issue is to be discussed in the following API/dev-review/TSC
```
Can you link to the GitHub Wiki pages where those meetings are described?

carlescufi commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,133 @@
+**********************
+
+=============  ===============================================================
+Labels         ``Hot Fix``, ``Trivial PR``, ``Maintainer``,
```
isn't the labe "Trivial" and not "Trivial PR"?

aescolar replied:
The plan was to rename Trivial => "Trivial PR" (To avoid developers assigning it to trivial bugs.

carlescufi replied:
Does it matter if we reuse the label for bugs and PRs?

aescolar replied:
If there is any purpose of using a "Trivial" label in a bug then we could use it. Today, I don't see any need for such.
But having too many labels with no clear purpose will add to the confusion (which we are trying to reduce with this PR)

carlescufi commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,133 @@
+*********************
+
+=============  ===============================================================
+Labels         ``priority:{high|medium|low}``
```
Is there somewhere with more information about what those categories mean? For example in the development process we should have how many of those we area allowed to have open before release

aescolar replied:
Not that I'm aware. But yes, I agree it is important to clarify it. As soon as we would, it should be linked from here.

carlescufi replied:
I thought we did, strange. @nashif ?

nashif replied:
Not sure this is documented anywhere beside TSC meeting minutes. Need to add.

carlescufi commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,133 @@
+================ ===============================================================
+
+Issue only labels
+=================
```
I think it's important to note that an issue needs to have one of the 4 following labels:

- Bug
- Question
- Enhancement
- Feature Request

marc-hb replied:
The only way to make that clear and simple is for them to share a common prefix, I suggest: `Type: bug`, `Type: Question`,... This will also group them in the list.

carlescufi reviewed (changes requested):
Nice work @aescolar thanks for this! I've added a few comments

mbolivar commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,138 @@
+###########################################
+
+Lots of GitHub issues and pull requests are being created daily.
+To help us manage them we have “labels”:
```
nit: this colon (":") seems like it should be removed

aescolar replied:
would a semicolon seem better?

mbolivar replied:
I don't think so. A semicolon would mean "to classify them by area, type, [...] attention" is an independent clause, which it is not (https://www.proofreadnow.com/blog/bid/71799/The-Semicolon-and-Independent-Clauses).

I don't think any punctuation marks are needed (or appropriate)

aescolar replied:
Thanks, done

mbolivar commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,138 @@
+fix it.
+
+This saves us all time when searching, reduces the chances of the PR or issue
+being forgotten, speeds up reviewing, reduces the changes of duplicates, etc.
```
I'm having a little trouble understanding what "reduces the changes of duplicates" means -- do you think you can clarify this?

aescolar replied:
would it be enough to say "reduces the changes of duplicate issue reports"?
(it is just a quick motivational paragraph, I would not it want to become too long)

mbolivar replied:
OK, so this is about avoiding dupes? How about "avoids duplicate issue reports"?

mbolivar commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,138 @@
+|``Waiting for response``| The Zephyr developers are waiting for the submitter |
+|                        | to respond to a question, or address an issue.      |
++------------------------+-----------------------------------------------------+
+|``RFC``                 | The author would like input from the community. For |
```
Should we consider deprecating this label in favor of using draft pull requests only? https://github.blog/2019-02-14-introducing-draft-pull-requests/

I see below that `This PR should not be merged (Do Not Merge).  For work in progress, GitHub "draft" PRs are preferred"` is listed next to DNM. I also see `In progress` is a label which can apply to PRs that seems very similar.

As a reader, I'm a bit confused between the following labels for pull requests, which all seem to be related:

 - RFC
- In progress
- DNM
- GitHub "Draft" PRs

Is there an opportunity to consolidate these into just one label (or perhaps just use draft PRs)?

aescolar replied:
Maybe, but not trivially right now.
(Note that I'm not trying to fix all issues related to our way of working with this PR)
I'm hoping to start by document what we have, and at the same time, clean what is trivial to clean.
If we start by going further in this same PR it will just balloon out of proportion.
=> If you don't mind, let's take that discussion separately from this PR.

mbolivar replied:
Sure

mbolivar reviewed (commented):
Looks nice, thanks a lot for doing this!

I have a minor grammar nit and then some questions about the labels for "incomplete" or non-mergeable PRs.

aescolar commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,139 @@
+****************************
+
+=============  ===============================================================
+Labels         ``API``, ``dev-review``, ``TSC``
```
@carlescufi: Should we separate the API label into a "area: API", and "API (meeting)", to clear a bit things?

carlescufi replied:
I don't think it's necessary. API is API, we can go through everything marked as API during the meeting

aescolar replied:
My thinking was that such a label can both help you to filter, but also the author of the PR, to know the PR is going to be discussed there.

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,139 @@
+****************************
+
+=============  ===============================================================
+Labels         ``API``, ``dev-review``, ``TSC``
```
There's an organized person in me that would like to have all labels with a prefix (e.g., ``area: API``, ``meetings: API``) so that the labels are grouped together (``meetings: API``, ``meetings: Dev-review``, ``meetings: TSC``), otherwise they don't sort together.

aescolar replied:
I agree :)

aescolar replied:
Would it be reasonable to, by now, have this PR start as documentation of what there actually is, and then open an issue with this different fixes? (and have them applied once agreed in a a following PR). I just don't want this PR to drag long.

dbkinder replied:
OK.  Maybe you can create an issue to track this (adding a prefix to all labels).

nashif replied:
dont like the meeting prefix, it does not have to be meeting, this is something that can always be brought into discussion, on slack, in email and other places.

dbkinder replied:
maybe ``review:`` then?

dbkinder commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,139 @@
+
+=============  ===============================================================
+Labels         ``Hot Fix``, ``Trivial PR``, ``Maintainer``,
+               ``Security Review``, ``TSC``
```
Is this the same ``TSC`` mentioned above?  If it's different, then a prefix would clarify :)

aescolar replied:
Today it is the same, I would be happy to split it in two, but Anas was not too keen on it.

dbkinder reviewed (approved):
we can merge this now to document how labels are used, and submit a future fix for resolving label organization and specific names...

nashif commented on doc/development_process/review_process.rst:
```diff
@@ -70,8 +74,8 @@ on the fix, severity, and availability of someone to review them (other than the
 
-Trivial
-+++++++
+Trivial PR
```
Please keep it as Trivial, no need for the PR addition.

nashif commented on doc/development_process/review_process.rst:
```diff
@@ -55,8 +59,8 @@ Workflow
 
-Categories/Labels
------------------
+Review time categories/labels
```
This is already indicated in the top section, it is about time, keep it general, TSC and Security and labels are not related to time.

aescolar replied:
Ok, I will revert. The reason for the change was due to the headings' levels not being to clear, and somebody maybe understanding this to be the general category/label section.

nashif commented on doc/development_process/review_process.rst:
```diff
@@ -14,13 +14,17 @@ GitHub is used to support the standard open source practice of submitting
 the code base.
 
+Pull requests should be appropriately :ref:`labeled<gh_labels>`,
+and link to any relevant :ref:`bug or feature tracking issues<issue_tracking>`.
```
and linked

aescolar replied:
thanks, indeed.

nashif commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,139 @@
+.. _gh_labels:
```
can this document be integrated into the existing issues document as a sub-section?

aescolar replied:
Today that mini-page content is a bit disconnected from this one (I guess you meant issues.rst ?)
You have a PR editing that one.. So after this is merged, feel free to take this content in if you feel it fits better there.

nashif commented on doc/development_process/gh_labels.rst:
```diff
@@ -0,0 +1,139 @@
+``DNM``          This PR should not be merged (Do Not Merge).
+                 For work in progress, GitHub "draft" PRs are preferred
+``Stale PR``     PR which seems abandoned, and requires attention by the author
+``Blocked``      Blocked by another PR or issue
```
And issue can block another issue, so this applies to more than just PRs

aescolar replied:
correct, moved

nashif commented on doc/development_process/index.rst:
```diff
@@ -15,6 +15,7 @@ Development Model
    api_lifecycle
    communication
    documentation
+   gh_labels
```
IMO this should not be a top level section, it should go under something else like issues or proposals or the dev process.

aescolar committed:
doc: Reorganize development model section

Renamed the page "review_process.rst" to "dev_env_and_tools.rst"
as the old name did not cover all its content.
Moved the labels page content into this
"Developemnt Environment and Tools" page.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
Compared to what people had reviewed:
Following [this request](https://github.com/zephyrproject-rtos/zephyr/pull/15054#pullrequestreview-224923911) from @nashif the labels section is now moved into the "Developemnt Environment and Tools" page.
Due to this, all the headings are now 1 level lower to align them with that page.

Apart from that the only other change, due to a new reference added in 15ca53dbbab, the equivalent new reference which had been added in this PR is now removed.
```diff
diff --git a/doc/development_process/dev_env_and_tools.rst b/doc/development_process/dev_env_and_tools.rst
index 43a626054f..c46965e0af 100644
--- a/doc/development_process/dev_env_and_tools.rst
+++ b/doc/development_process/dev_env_and_tools.rst
@@ -15,7 +15,7 @@ patches, which are then reviewed by the project members before being applied to
 the code base.
 
 Pull requests should be appropriately :ref:`labeled<gh_labels>`,
-and linked to any relevant :ref:`bug or feature tracking issues<issue_tracking>`
+and linked to any relevant :ref:`bug or feature tracking issues<bug_reporting>`
 .
 
 The Zephyr project uses GitHub for code reviews and Git tree management. When
diff --git a/doc/development_process/issues.rst b/doc/development_process/issues.rst
index 16552f1a49..ceca4d095b 100644
--- a/doc/development_process/issues.rst
+++ b/doc/development_process/issues.rst
@@ -1,5 +1,3 @@
-.. _issue_tracking:
-
 .. _bug_reporting:
 
 Bug Reporting
```
